File: frontend/src/services/modules.ts
import { api } from './api';

// Generic CRUD for project modules
function createModuleService<T, C>(basePath: string) {
  return {
    list: (params?: Record<string, string>) => {
      const qs = params ? '?' + new URLSearchParams(params).toString() : '';
      return api.get<T[]>(`/${basePath}${qs}`);
    },
    get: (id: number) => api.get<T>(`/${basePath}/${id}`),
    create: (projectId: number, data: C) =>
      api.post<T>(`/${basePath}?project_id=${projectId}`, data),
    update: (id: number, data: Partial<T>) => api.patch<T>(`/${basePath}/${id}`, data),
    delete: (id: number) => api.delete(`/${basePath}/${id}`),
  };
}

export interface Risk {
  id: number;
  folio: string;
  title: string;
  description?: string;
  category?: string;
  probability: number;
  impact: number;
  severity: number;
  mitigation_strategy?: string;
  status: string;
  identification_date?: string;
  deadline?: string;
  project_id: number;
  responsible_id?: number;
  created_at: string;
}

export interface Issue {
  id: number;
  folio: string;
  title: string;
  description?: string;
  type: string;
  priority: string;
  status: string;
  resolution?: string;
  report_date?: string;
  commitment_date?: string;
  project_id: number;
  responsible_id?: number;
  created_at: string;
}

export interface Change {
  id: number;
  folio: string;
  title: string;
  description?: string;
  change_type: string;
  impact?: string;
  requested_by?: string;
  request_date?: string;
  status: string;
  comments?: string;
  project_id: number;
  created_at: string;
}

export interface Document {
  id: number;
  folio: string;
  name: string;
  description?: string;
  category: string;
  file_path?: string;
  file_type?: string;
  file_size?: number;
  version?: string;
  project_id: number;
  created_at: string;
}

export interface Lesson {
  id: number;
  folio: string;
  title: string;
  description?: string;
  category: string;
  project_phase?: string;
  recommendation?: string;
  project_id: number;
  created_at: string;
}

export interface Minute {
  id: number;
  folio: string;
  title: string;
  meeting_date: string;
  participants?: string;
  topics?: string;
  agreements?: string;
  source: string;
  ai_model_used?: string;
  project_id: number;
  created_at: string;
}

export const risksService = createModuleService<Risk, Partial<Risk>>('risks');
export const issuesService = createModuleService<Issue, Partial<Issue>>('issues');
export const changesService = createModuleService<Change, Partial<Change>>('changes');
export const documentsService = createModuleService<Document, Partial<Document>>('documents');
export const lessonsService = createModuleService<Lesson, Partial<Lesson>>('lessons');
export const minutesService = {
  ...createModuleService<Minute, Partial<Minute>>('minutes'),
  generate: (data: {
    transcript: string;
    project_id: number;
    title: string;
    meeting_date: string;
    language?: string;
  }) =>
    api.post<{ minute: Minute; generated_text: string; model_used: string }>(
      '/minutes/generate',
      data,
    ),
};

File: backend/app/models/modules.py
from sqlalchemy import Column, String, Integer, Date, Text, ForeignKey
from sqlalchemy.orm import relationship

from app.database import Base
from app.models.base import TimestampMixin
from app.models.mixins import TenantScopedMixin


class Risk(TenantScopedMixin, TimestampMixin, Base):
    __tablename__ = "risks"

    folio = Column(String(50), unique=True, nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    category = Column(String(100), nullable=True)
    probability = Column(Integer, nullable=False)     # 1-5
    impact = Column(Integer, nullable=False)           # 1-5
    severity = Column(Integer, nullable=False)         # probability * impact
    mitigation_strategy = Column(Text, nullable=True)
    status = Column(String(50), default="open")        # open, mitigated, closed
    identification_date = Column(Date, nullable=False)
    deadline = Column(Date, nullable=True)

    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    responsible_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    project = relationship("Project", back_populates="risks")


class Issue(TenantScopedMixin, TimestampMixin, Base):
    __tablename__ = "issues"

    folio = Column(String(50), unique=True, nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    type = Column(String(50), nullable=False)          # action, issue, decision (AID)
    priority = Column(String(20), nullable=False)      # Alta, Media, Baja
    status = Column(String(50), default="open")        # open, in_progress, resolved, closed
    resolution = Column(Text, nullable=True)
    report_date = Column(Date, nullable=False)
    commitment_date = Column(Date, nullable=True)

    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    responsible_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    project = relationship("Project", back_populates="issues")


class Change(TenantScopedMixin, TimestampMixin, Base):
    __tablename__ = "changes"

    folio = Column(String(50), unique=True, nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    change_type = Column(String(50), nullable=False)   # scope, time, cost, resource
    impact = Column(Text, nullable=True)
    requested_by = Column(String(255), nullable=True)
    request_date = Column(Date, nullable=False)
    status = Column(String(50), default="in_review")   # in_review, approved, rejected, implemented
    approved_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    approval_date = Column(Date, nullable=True)
    comments = Column(Text, nullable=True)

    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    project = relationship("Project", back_populates="changes")


class Document(TenantScopedMixin, TimestampMixin, Base):
    __tablename__ = "documents"

    folio = Column(String(50), unique=True, nullable=False, index=True)
    name = Column(String(255), nullable=False)
    description = Column(String(500), nullable=True)
    category = Column(String(100), nullable=True)      # plan, report, contract, other
    file_path = Column(String(500), nullable=False)
    file_type = Column(String(20), nullable=True)
    file_size = Column(Integer, nullable=True)          # bytes
    version = Column(Integer, default=1)

    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    uploaded_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    project = relationship("Project", back_populates="documents")


class Lesson(TenantScopedMixin, TimestampMixin, Base):
    __tablename__ = "lessons"

    folio = Column(String(50), unique=True, nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    category = Column(String(50), nullable=True)       # success, improvement, error
    project_phase = Column(String(100), nullable=True)
    recommendation = Column(Text, nullable=True)

    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    recorded_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    project = relationship("Project", back_populates="lessons")


class Minute(TenantScopedMixin, TimestampMixin, Base):
    __tablename__ = "minutes"

    folio = Column(String(50), unique=True, nullable=False, index=True)
    title = Column(String(255), nullable=False)
    meeting_date = Column(Date, nullable=False)
    participants = Column(Text, nullable=True)         # JSON array of names/ids
    topics = Column(Text, nullable=True)
    agreements = Column(Text, nullable=True)
    next_meeting_date = Column(Date, nullable=True)
    source = Column(String(50), default="manual")      # manual, ai_generated

    # AI fields
    transcript_text = Column(Text, nullable=True)
    ai_model_used = Column(String(100), nullable=True)
    ai_generation_time_ms = Column(Integer, nullable=True)

    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)
    recorded_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    project = relationship("Project", back_populates="minutes")
